kroxylicious/src/main/java/io/kroxylicious/proxy/internal/FilterHandler.java

```diff
@@ -112,66 +111,163 @@ else if (requestFilterResult.message() != null) {
-                        filterContext.closeConnection();
-                        return;
-                    }
-                    if (responseFilterResult.drop()) {
-                        if (LOGGER.isDebugEnabled()) {
-                            LOGGER.debug("{}: Filter{} drops {} response",
-                                    ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
-                        }
-                    }
-                    else if (responseFilterResult.message() != null) {
-                        ResponseHeaderData header = responseFilterResult.header() == null ? decodedFrame.header() : (ResponseHeaderData) responseFilterResult.header();
-                        filterContext.forwardResponseInternal(header, responseFilterResult.message());
-                    }
-                    if (responseFilterResult.closeConnection()) {
-                        filterContext.closeConnection();
-                    }
-
-                });
-
+    private CompletableFuture<Void> doRead(ChannelHandlerContext ctx, Object msg) {
+        if (msg instanceof DecodedResponseFrame<?> decodedFrame) {
+            var filterContext = new DefaultFilterContext(filter, ctx, decodedFrame, null, timeoutMs, sniHostname);
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug("{}: Dispatching upstream {} response to filter {}: {}",
+                        ctx.channel(), decodedFrame.apiKey(), filterDescriptor(), msg);
             }
+            var stage = invoker.onResponse(decodedFrame.apiKey(), decodedFrame.apiVersion(),
+                    decodedFrame.header(), decodedFrame.body(), filterContext);
+            if (stage == null) {
+                LOGGER.error("{}: Filter{} for {} response unexpectedly returned null. This is a coding error in the filter. Closing connection.",
+                        ctx.channel(), filterDescriptor(), decodedFrame.apiKey());
+                filterContext.closeConnection();
+                return CompletableFuture.completedFuture(null);
+            }
+            var future = stage.toCompletableFuture();
+            boolean defer = !future.isDone();
+            var maybeDeferred = defer ? handleDeferredStage(ctx, future, decodedFrame) : future;
+            var execute = executeRead(ctx, decodedFrame, filterContext, maybeDeferred);
+            var maybeUndeferred = defer ? handleUndeferRead(execute) : execute;
+            return maybeUndeferred.thenApply(responseFilterResult -> null);
         }
         else {
             if (!(msg instanceof OpaqueResponseFrame)) {
                 LOGGER.warn("Unexpected message reading from upstream: {}", msg, new IllegalStateException());
             }
             ctx.fireChannelRead(msg);
+            return CompletableFuture.completedFuture(null);
         }
     }
 
+    private CompletableFuture<?> handleUndeferRead(CompletableFuture<?> execute) {
```

Comment:
I'm not sure about the naming. The method `handleUndefer` had me scratching my head. How about `handleDeferredStageCompletion`